### ISSUE TITLE
Encryption of chat massages and rcon password

### ISSUE THREAD

**Issue by Zwelf** [teeworlds/teeworlds#1379]
My suggestion for encryption for teeworlds:

Every client and server generate a key pair when they start.
The master servers also generate a key pair which is signed by the master key.
The master key is saved in every client and server.
The server sends in the registration his public key (encrypted, befor it sends its registration it requests the master server's key and check if it is valid).

When a client requests the server list, it first requests the key of the master server and checks if it is valid, then the master server sends the list signed, every server with its key.
When a client then joins a server and if it doesn't know the key from the server list (e.g. lan servers) it requests it from the server.
Then the client sends its key (encrypted and has to be validate from the server).
Then the client sends another key (random), but this time a symmetric encryption method (faster), the server sends back the key encrypted and verified, and everything will be encrypted with this key.

All chat massages and sended rcon password were be encrypted

Tell what you think about this methode.

**Comment by KillaBilla** [comment]
Encrypting the server and rcon password seems like a good idea but using a one-way hash function (#1209) might be better since it does not waste as much bandwidth.

IMHO encrypting the chat isn't really necessary as confident chatter should not take place in teeworlds.

Too much bandwidth is wasted if the master server has the public key for every server. Having an additional 4096 bit for the public key (if you want decent security) increases the data size by more than 2500%.

**Comment by Zwelf** [comment]
It doesn't mean that it is bad to encrypt chat messages if no confident chatter takes place in teeworlds.
It doesn't hurt encrypting chat messages.

**Comment by PierreSassoulas** [comment]
I did'nt understand what you mean when you say it does not hurt. If the data teeworld need to exchange are heavier because of encryption, it means encryption has a cost, rigth ?

**Comment by Zwelf** [comment]
Right it has costs, but when you encrypt, then additionally encrypting chat messages doesn't need much resources.

Idea for less bandwidth: The server sends only ip-addresses first. But when a client wants to join a server, it requests the key. Drawback: The master server can track the servers which you join.

**Comment by KillaBilla** [comment]
@Zwelf The master server can't track which servers you join.

**Comment by Zwelf** [comment]
Yes I know this. I presented an idea, with encryption and less traffic than sending all public keys.
> when a client wants to join a server, it requests the key.

I want to say that the client requests the key from the master server. __This way__ the there is less traffic the master server knows which server you join.

**Comment by Henningstone** [comment]
Less traffic... not necessarily. There are a lot clients who enter servers every minute, and sending all time this key means even much more traffic for the masters then only sending the key to registering servers? Or did I get something wrong?

**Comment by Zwelf** [comment]
To prevent man in the middle attacks, the client has to know the public key before it joins the server.
First I was thinking about sending all keys, when a client requests the server list. This takes much bandwidth.
To reduce the traffic, the server sends only the ip list as always, but the client requests the key of the server it wants to join. Then the master server doesn't have to send 900x public keys, but rather it sends one public key to each client.
This way the master server could stay as it is now and there could be additional key servers.

**Comment by BotoX** [comment]
Transmitting the rcon password salted with the current unix time and sha512 hashed seems like a good idea, no?

I don't like the idea of adding too much complexity to the system with your encryption model, what if the masters go down? Also have to counter reflection attacks, etc.

**Comment by Zwelf** [comment]
What if the master server goes down?: The client register at the next master server (as always).
Counter reflection attacks: #986 handle this.
